File: packages/go/hushspec/bundle_vectors_test.go
package hushspec

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"gopkg.in/yaml.v3"
)

// The normative policy-bundle vectors (spec/hushspec-bundle.md section 7,
// fixtures/bundle/). An implementation conforms as a bundle verifier exactly
// when every case here returns the expected outcome, so this runner is the Go
// SDK's conformance statement for bundle verification.

// bundleVectorsVersion is the manifest version this runner understands.
const bundleVectorsVersion = "0.1.0"

// bundleVectorCaseCount is the number of cases the specification's table
// enumerates. Pinning it means a vector added upstream cannot be silently
// skipped by a runner that only iterates what it finds.
const bundleVectorCaseCount = 8

type bundleVectorFile struct {
	Version     string               `yaml:"hushspec_bundle_vectors"`
	Description string               `yaml:"description"`
	Defaults    bundleVectorDefaults `yaml:"defaults"`
	Cases       []bundleVectorCase   `yaml:"cases"`
}

type bundleVectorDefaults struct {
	Keyring string `yaml:"keyring"`
	Now     string `yaml:"now"`
}

type bundleVectorCase struct {
	Name    string    `yaml:"name"`
	Bundle  string    `yaml:"bundle"`
	Keyring string    `yaml:"keyring"`
	Policy  string    `yaml:"policy"`
	Now     string    `yaml:"now"`
	Expect  yaml.Node `yaml:"expect"`
	Note    string    `yaml:"note"`
}

// expectation reads the manifest's untagged `expect`: the scalar `valid`, or
// a mapping `{invalid: <reason code>}`.
func (c *bundleVectorCase) expectation(t *testing.T) string {
	t.Helper()
	var scalar string
	if err := c.Expect.Decode(&scalar); err == nil {
		if scalar != "valid" {
			t.Fatalf("%s: expect %q is neither \"valid\" nor an invalid mapping", c.Name, scalar)
		}
		return "valid"
	}
	var invalid struct {
		Invalid string `yaml:"invalid"`
	}
	if err := c.Expect.Decode(&invalid); err != nil || invalid.Invalid == "" {
		t.Fatalf("%s: expect is not {invalid: <reason code>}", c.Name)
	}
	return invalid.Invalid
}

func loadBundleVectors(t *testing.T) (string, *bundleVectorFile) {
	t.Helper()
	root := filepath.Join(fixtureRepoRoot(t), "fixtures", "bundle")
	raw, err := os.ReadFile(filepath.Join(root, "vectors.yaml"))
	if err != nil {
		t.Fatalf("failed to read the bundle vector manifest: %v", err)
	}
	var manifest bundleVectorFile
	if err := yaml.Unmarshal(raw, &manifest); err != nil {
		t.Fatalf("failed to parse the bundle vector manifest: %v", err)
	}
	if manifest.Version != bundleVectorsVersion {
		t.Fatalf("manifest version %q, this runner understands %q",
			manifest.Version, bundleVectorsVersion)
	}
	if len(manifest.Cases) != bundleVectorCaseCount {
		t.Fatalf("bundle spec 7 publishes %d vectors, the manifest lists %d",
			bundleVectorCaseCount, len(manifest.Cases))
	}
	return root, &manifest
}

// TestBundleVectors walks fixtures/bundle/vectors.yaml: every case must return
// its expected outcome, valid or the expected reason code of bundle spec 5.4.
func TestBundleVectors(t *testing.T) {
	root, manifest := loadBundleVectors(t)

	for _, tc := range manifest.Cases {
		t.Run(tc.Name, func(t *testing.T) {
			keyringPath := tc.Keyring
			if keyringPath == "" {
				keyringPath = manifest.Defaults.Keyring
			}
			keyringJSON, err := os.ReadFile(filepath.Join(root, keyringPath))
			if err != nil {
				t.Fatalf("failed to read the keyring: %v", err)
			}
			keyring, err := LoadKeyring(keyringJSON)
			if err != nil {
				t.Fatalf("the keyring did not load: %v", err)
			}

			nowText := tc.Now
			if nowText == "" {
				nowText = manifest.Defaults.Now
			}
			now, err := time.Parse(time.RFC3339, nowText)
			if err != nil {
				t.Fatalf("%q is not a timestamp: %v", nowText, err)
			}

			bundle, err := os.ReadFile(filepath.Join(root, tc.Bundle))
			if err != nil {
				t.Fatalf("failed to read the bundle: %v", err)
			}

			options := VerifyBundleOptions{Keyring: keyring, Now: now}
			if tc.Policy != "" {
				// Every vector that names a policy names one that resolves, so a
				// resolution failure is a broken fixture. Treating it as check 4's
				// own failure would let a vector expecting policy_mismatch pass
				// without the verifier ever running.
				resolution, resolveErr := ResolveFileWithOptions(
					filepath.Join(root, tc.Policy), ResolveOptions{})
				if resolveErr != nil {
					t.Fatalf("the vector names a policy that does not resolve: %v", resolveErr)
				}
				options.PolicyResolution = resolution
			}

			result := VerifyBundle(bundle, options)
			actual := result.Reason
			if result.OK {
				actual = "valid"
			}

			want := tc.expectation(t)
			if actual != want {
				t.Errorf("expected %s, got %s (%s)", want, actual, result.Detail)
			}
			if result.OK != (want == "valid") {
				t.Errorf("OK=%v but the expected outcome is %q", result.OK, want)
			}
			if result.OK && result.Reason != "" {
				t.Errorf("a valid bundle carries no reason code, got %q", result.Reason)
			}
			if want == "valid" {
				if len(result.KeyIDs) == 0 {
					t.Error("a valid bundle names at least one verifying key")
				}
				if result.PolicyChecked != (tc.Policy != "") {
					t.Errorf("PolicyChecked=%v for policy %q", result.PolicyChecked, tc.Policy)
				}
				if result.VerifiedAt != "2026-09-15T12:00:00.000Z" {
					t.Errorf("verified_at should stamp the verifier's clock, got %q", result.VerifiedAt)
				}
			}
		})
	}
}

// TestBundleVectorsUseTheClosedReasonSet keeps the manifest and this SDK on
// one vocabulary: a typo in a vector cannot pass as a new code, and every code
// the specification defines is exercised by some vector.
func TestBundleVectorsUseTheClosedReasonSet(t *testing.T) {
	_, manifest := loadBundleVectors(t)

	known := map[string]bool{}
	for _, reason := range BundleReasons {
		known[reason] = true
	}
	covered := map[string]bool{}
	for _, tc := range manifest.Cases {
		expected := tc.expectation(t)
		if expected == "valid" {
			continue
		}
		if !known[expected] {
			t.Errorf("%s: %q is not a bundle spec 5.4 reason code", tc.Name, expected)
		}
		covered[expected] = true
	}
	for _, reason := range BundleReasons {
		if !covered[reason] {
			t.Errorf("no vector expects %s", reason)
		}
	}
}

// TestBundlesValidateAgainstTheSchema checks every published bundle against
// schemas/hushspec-bundle.v0.schema.json: the envelope against the root, and
// the decoded payload against $defs/Statement. This SDK carries no JSON Schema
// engine, so the constraints the schema states for the members it reads are
// asserted directly -- the three `const` members are read out of the schema
// itself, so a bundle, a Go constant and the published schema cannot drift
// apart unnoticed.
func TestBundlesValidateAgainstTheSchema(t *testing.T) {
	root, manifest := loadBundleVectors(t)
	schemaPath := filepath.Join(fixtureRepoRoot(t), "schemas", "hushspec-bundle.v0.schema.json")
	schemaJSON, err := os.ReadFile(schemaPath)
	if err != nil {
		t.Fatalf("the bundle schema is not published: %v", err)
	}
	var schema map[string]any
	if err := json.Unmarshal(schemaJSON, &schema); err != nil {
		t.Fatalf("the bundle schema is not JSON: %v", err)
	}

	// The constants this SDK compares against are the schema's own, so a
	// divergence between the two is a failure here rather than a silent
	// disagreement with every other implementation.
	payloadType := schemaConst(t, schema, "properties", "payloadType")
	statementType := schemaConst(t, schema, "$defs", "Statement", "properties", "_type")
	predicateType := schemaConst(t, schema, "$defs", "Statement", "properties", "predicateType")
	for _, pin := range []struct{ schema, sdk string }{
		{payloadType, BundlePayloadType},
		{statementType, BundleStatementType},
		{predicateType, BundlePredicateType},
	} {
		if pin.schema != pin.sdk {
			t.Fatalf("the schema pins %q where this SDK uses %q", pin.schema, pin.sdk)
		}
	}

	seen := map[string]bool{}
	for _, tc := range manifest.Cases {
		if seen[tc.Bundle] {
			continue
		}
		seen[tc.Bundle] = true
		t.Run(tc.Bundle, func(t *testing.T) {
			data, err := os.ReadFile(filepath.Join(root, tc.Bundle))
			if err != nil {
				t.Fatalf("failed to read the bundle: %v", err)
			}
			envelope, err := ParseBundle(data)
			if err != nil {
				t.Fatalf("every published bundle is a well-formed envelope: %v", err)
			}
			if envelope.PayloadType != payloadType {
				t.Errorf("payloadType %q, the schema pins %q", envelope.PayloadType, payloadType)
			}
			payload, err := envelope.PayloadBytes()
			if err != nil {
				t.Fatalf("the payload does not decode: %v", err)
			}
			// malformed-predicate-type is deliberately a statement this build
			// rejects, so only its envelope half is asserted here.
			var statement BundleStatement
			if err := strictUnmarshalJSON(payload, &statement); err != nil {
				t.Fatalf("the payload is not a statement: %v", err)
			}
			if len(statement.Subject) != 1 {
				t.Errorf("a bundle attests exactly one subject, found %d", len(statement.Subject))
			}
			if statement.Type != statementType {
				t.Errorf("_type %q, the schema pins %q", statement.Type, statementType)
			}
			if len(statement.Predicate.Chain) == 0 {
				t.Error("the chain must hold at least one link")
			}
		})
	}
}

// schemaConst reads the `const` of the schema member at path, failing when the
// path does not lead to one.
func schemaConst(t *testing.T, schema map[string]any, path ...string) string {
	t.Helper()
	node := schema
	for _, segment := range path {
		next, ok := node[segment].(map[string]any)
		if !ok {
			t.Fatalf("the bundle schema has no object at %s", strings.Join(path, "."))
		}
		node = next
	}
	value, ok := node["const"].(string)
	if !ok {
		t.Fatalf("the bundle schema member at %s has no string const", strings.Join(path, "."))
	}
	return value
}

File: packages/hushspec/tests/bundle-vectors.test.ts
import { existsSync, readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import YAML from 'yaml';
import { describe, expect, it } from 'vitest';
import {
  BUNDLE_REASONS,
  bundleStatement,
  pae,
  parseBundle,
  verifyBundle,
  type BundleReason,
} from '../src/bundle.js';
import { loadKeyring } from '../src/signing.js';
import { createCompositeLoader, resolveWithOptions } from '../src/resolve.js';
import type { Resolution } from '../src/resolve.js';
import { parse } from '../src/parse.js';

/**
 * The normative policy-bundle vectors (bundle spec 7, `fixtures/bundle/`).
 *
 * An implementation conforms as a bundle verifier if, for every case in
 * `vectors.yaml`, it returns the expected outcome: `valid`, or invalid with
 * the expected reason code of bundle spec 5.4.
 */

const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../..');
const fixturesRoot = path.join(repoRoot, 'fixtures', 'bundle');

interface Manifest {
  hushspec_bundle_vectors: string;
  description?: string;
  defaults: { keyring: string; now: string };
  cases: Case[];
}

interface Case {
  name: string;
  bundle: string;
  keyring?: string;
  policy?: string;
  now?: string;
  expect: 'valid' | { invalid: string };
  note?: string;
}

const manifest = YAML.parse(
  readFileSync(path.join(fixturesRoot, 'vectors.yaml'), 'utf8'),
) as Manifest;

/** Resolve a policy the way `h2h bundle verify --policy` does. */
function resolvePolicy(filePath: string): Resolution | undefined {
  if (!existsSync(filePath)) return undefined;
  const parsed = parse(readFileSync(filePath, 'utf8'));
  if (!parsed.ok) return undefined;
  try {
    return resolveWithOptions(parsed.value, {
      source: filePath,
      loader: createCompositeLoader(),
    });
  } catch {
    // A policy that will not resolve has nothing to compare, which is check
    // 4's own failure -- never a thrown error out of the runner.
    return undefined;
  }
}

describe('policy bundle vectors', () => {
  it('declares the manifest version the runner understands', () => {
    expect(manifest.hushspec_bundle_vectors).toBe('0.1.0');
    expect(manifest.cases.length).toBeGreaterThan(0);
  });

  for (const testCase of manifest.cases) {
    it(`${testCase.name}`, () => {
      const keyring = loadKeyring(
        readFileSync(
          path.join(fixturesRoot, testCase.keyring ?? manifest.defaults.keyring),
          'utf8',
        ),
      );
      const bundleJson = readFileSync(path.join(fixturesRoot, testCase.bundle), 'utf8');
      const policy = testCase.policy == null
        ? undefined
        : resolvePolicy(path.join(fixturesRoot, testCase.policy));

      const outcome = verifyBundle(bundleJson, {
        keyring,
        now: testCase.now ?? manifest.defaults.now,
        // A policy the manifest names but that does not resolve is still
        // check 4's input: `verifyBundle` reports `policy_mismatch` for it.
        ...(testCase.policy == null ? {} : { policy: policy ?? { hushspec: '0.0.0' } }),
      });

      if (testCase.expect === 'valid') {
        expect(outcome.ok ? 'valid' : `${outcome.reason}: ${outcome.detail}`).toBe('valid');
        if (!outcome.ok) return;
        expect(outcome.keyIds.length).toBeGreaterThan(0);
        expect(outcome.contentHash).toMatch(/^sha256:[0-9a-f]{64}$/);
        expect(outcome.policyChecked).toBe(testCase.policy != null);
        expect(outcome.verifiedAt).toBe(testCase.now ?? manifest.defaults.now);
      } else {
        expect(outcome.ok).toBe(false);
        if (outcome.ok) return;
        expect(outcome.reason).toBe(testCase.expect.invalid as BundleReason);
        expect(outcome.detail.length).toBeGreaterThan(0);
      }
    });
  }

  it('covers every reason code the specification defines', () => {
    const expected = new Set(
      manifest.cases
        .map(testCase => (testCase.expect === 'valid' ? undefined : testCase.expect.invalid))
        .filter((code): code is string => code != null),
    );
    for (const code of BUNDLE_REASONS) {
      expect(expected.has(code), `no vector expects ${code}`).toBe(true);
    }
  });
});

describe('bundle parsing', () => {
  const validBundle = (): string =>
    readFileSync(path.join(fixturesRoot, 'bundles/valid.bundle.json'), 'utf8');

  it('parses an envelope and reads its statement', () => {
    const envelope = parseBundle(validBundle());
    expect(envelope.payloadType).toBe('application/vnd.in-toto+json');
    expect(envelope.signatures).toHaveLength(1);

    const statement = bundleStatement(envelope);
    expect(statement._type).toBe('https://in-toto.io/Statement/v1');
    expect(statement.predicateType).toBe(
      'https://hushspec.dev/attestation/policy-bundle/v0.1',
    );
    expect(statement.predicate.bundle_version).toBe('0.1');
    expect(statement.subject).toHaveLength(1);
  });

  it('refuses a document that is not a DSSE envelope', () => {
    expect(() => parseBundle('{}')).toThrow(/payloadType is required/);
    expect(() => parseBundle('not json')).toThrow(/not JSON/);
    expect(() => parseBundle(JSON.parse(validBundle()) as unknown)).not.toThrow();
  });

  it('builds the PAE the specification quotes', () => {
    // The DSSE specification's own example.
    expect(pae('http://example.com/HelloWorld', Buffer.from('hello world')).toString('utf8'))
      .toBe('DSSEv1 29 http://example.com/HelloWorld 11 hello world');
    // Lengths are byte counts, not character counts.
    expect(pae('t', Buffer.from('é', 'utf8')).toString('utf8')).toBe('DSSEv1 1 t 2 é');
    // Every HushSpec bundle carries the same 28-byte type.
    expect(pae('application/vnd.in-toto+json', Buffer.from('{}')).toString('utf8'))
      .toBe('DSSEv1 28 application/vnd.in-toto+json 2 {}');
  });

  it('needs exactly one key source', () => {
    expect(() => verifyBundle(validBundle(), {})).toThrow(/keyring or a publicKeyPem/);
    expect(() =>
      verifyBundle(validBundle(), { keyring: '{}', publicKeyPem: 'x' })).toThrow(/not both/);
  });

  it('verifies under a single public key as well as a keyring', () => {
    const publicKeyPem = readFileSync(
      path.join(repoRoot, 'fixtures/signing/keys/test-signing.pub.pem'),
      'utf8',
    );
    const outcome = verifyBundle(validBundle(), { publicKeyPem, now: manifest.defaults.now });
    expect(outcome.ok).toBe(true);
  });
});
